hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreEngine.java

```diff
@@ -229,12 +232,23 @@ public HStoreFile createStoreFileAndReader(StoreFileInfo info) throws IOExceptio
   /**
    * Validates a store file by opening and closing it. In HFileV2 this should not be an expensive
    * operation.
-   * @param path the path to the store file
+   * @param path         the path to the store file
+   * @param isCompaction whether this is called from the context of a compaction
    */
-  public void validateStoreFile(Path path) throws IOException {
+  public void validateStoreFile(Path path, boolean isCompaction) throws IOException {
     HStoreFile storeFile = null;
     try {
       storeFile = createStoreFileAndReader(path);
+      if (conf.getBoolean(READ_FULLY_ON_VALIDATE_KEY, DEFAULT_READ_FULLY_ON_VALIDATE)) {
```

Comment:
This is the core of the new feature. Anywhere that we previously would invoke `validateStoreFile()` will now conditionally include the more rigorous check. There are a couple places where we do NOT call `validateStoreFile` after writing it, I'm not yet sure why.